[Dave Low] iP

davelowqx opened the pull request:
## DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://google.com))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://google.com).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Level-1

Commit:
Level-1: Greet, Echo, Exit

Commit:
Level-2

Commit:
Level-2: Add, List

Commit:
Level-3

Commit:
Level-3: Mark as Done

Commit:
Level-4: ToDo, Event, Deadline

Commit:
Level-4

Commit:
A-TextUiTesting

Commit:
A-TextUiTesting: Automated Text UI Testing

Commit:
Level-5

Commit:
Level-5: Handle Errors

Commit:
Level-6: delete

Commit:
Level-6: delete

Commit:
Merge branch 'master' of https://github.com/davelowqx/ip

Commit:
minor optimisations

Commit:
REMOVE: empty function

Commit:
formatting

Commit:
removed text output from commit

Commit:
removed localdb from commit

Commit:
Level-8: Dates and Time

jovyntls commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,12 @@
+public class Deadline extends Task {
+    String by;
+    public Deadline(String desc, String by) throws Exception{
```
```suggestion
    String by;
    
    public Deadline(String desc, String by) throws Exception{
```

HongshuW replied:
In addition to this change, don't forget to add white space in the method definition.
```suggestion
    String by;
    
    public Deadline(String desc, String by) throws Exception {
```

jovyntls commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,69 @@
+                default: // changes made
+                    if (input.matches("done \\d+")) {
+                        int i = Integer.parseInt(input.substring("done ".length()));
+                        System.out.println("Nice! I've marked this task as done:");
+                        Task t = tasks.get(i - 1);
+                        t.markAsComplete();
+                        System.out.println(t);
+                    } else if (input.matches("delete \\d+")) {
+                        int i = Integer.parseInt(input.substring("delete ".length()));
+                        System.out.println("Noted. I've removed this task:");
+                        Task t = tasks.remove(i - 1);
+                        System.out.println(t);
+                    } else {
+                        try {
+                            Task t;
+                            if (input.matches("event.*")) {
+                                int k = input.indexOf("/at");
+                                if (k < 0) {
+                                    throw new DukeException.MissingArgumentException("/at");
+                                }
+                                t = new Event(input.substring("event".length(), k).trim(), input.substring(k + 3).trim());
+                            } else if (input.matches("deadline.*")) {
+                                int k = input.indexOf("/by");
+                                if (k < 0) {
+                                    throw new DukeException.MissingArgumentException("/by");
+                                }
+                                t = new Deadline(input.substring("deadline".length(), k).trim(), input.substring(k + 3).trim());
+                            } else if (input.matches("todo.*")) {
+                                t = new Todo(input.substring("todo".length()));
+                            } else {
+                                throw new DukeException.UnknownInputException();
+                            }
+                            tasks.add(t);
+                            System.out.printf("Got it. I've added this task:\n%s\nNow you have %d tasks in the list\n", t, tasks.size());
+                        } catch (Exception err) {
+                            System.out.println(err.getMessage());
+                        }
+                    }
+            }
+        }
```
Perhaps you could extract some of the logic out into a method to reduce the amount of indentation?

jovyntls commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,69 @@
+                        System.out.printf("%d.%s\n", i + 1, tasks.get(i));
+                    }
+                    break;
+                default: // changes made
```
The default block in this switch expression is a bit long and confusing. Would it be more readable if you used one case for each command?

jovyntls reviewed (commented):
I like the separation of DukeException into subclasses as it makes the type of exception clearer and more specific! I also like how you used regex in matching commands. 😄

HongshuW commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,21 @@
+public abstract class DukeException extends Exception {
+    public DukeException(String s) {
```
Perhaps you could use a more intuitive variable name, such as errorMessage, instead of s.

HongshuW commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,12 @@
+
+    @Override
+    public String toString() {
+        return String.format("[E]%s (at: %s)", super.toString(), this.at);
```
I like how you used String.format() here instead of concatenation.

HongshuW commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,69 @@
+        while (true) {
+            input = scanner.nextLine();
+            switch (input) {
+                case "bye":
```
Should the case clauses have indentation?

HongshuW reviewed (commented):
LGTM. Maybe double check whether the program follows coding standard before merging?

Commit:
Merge branch 'branch-Level-7'

Commit:
Level-8: Dates and Times

Commit:
nest classes Event, Deadline, Todo in Task

Commit:
A-MoreOOP: Use More OOP

Commit:
A-Packages: Organize into Packages

jiayi1129 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,226 @@
+        return;
+    }
+
+    public class Storage {
```
Perhaps you can consider abstracting out these classes such as `Storage`, `Ui`, `Parser`, `Tasklist` into their own individual files/packages? Let me know what you think!

jiayi1129 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,28 @@
+package duke;
+
+public abstract class DukeException extends Exception {
+    public DukeException(String s) {
+        super("☹ OOPS!!! " + s);
+    }
+
+    public static class EmptyDescriptionException extends DukeException {
+        public EmptyDescriptionException() {
+            super("The description of a task cannot be empty.");
+        }
+    }
+    public static class UnknownInputException extends DukeException {
+        public UnknownInputException() {
+            super("I'm sorry, but I don't know what that means :-(");
+        }
+    }
+    public static class MissingArgumentException extends DukeException {
+        public MissingArgumentException(String arg) {
+            super("Missing argument: " +  arg);
+        }
+    }
+    public static class ParseException extends DukeException {
+        public ParseException(String s) {
+            super("Error reading line:" + s);
+        }
+    }
+}
```
I really like how you created these various types of `DukeException`! Great work! 👍

jiayi1129 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,126 @@
+    };
+
+    public static Task fromRepr(String s) throws DukeException {
+        if (!s.matches("[TED][|][01][|].+[|].*")) {
```
Perhaps you can create a private static final variable to hold the magic literal for the regex expression?

jiayi1129 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,126 @@
+        if (!s.matches("[TED][|][01][|].+[|].*")) {
+            throw new DukeException.ParseException(s);
+        }
+        String[] arr = s.split("[|]");
```
Perhaps you can consider creating a private static final variable to hold this magic literal for the delimiter split as well? I notice this issue in other places as well.

jiayi1129 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,226 @@
+        }
+        this.ui.init();
+        boolean b;
+        while ((b = this.parser.parseCommand(this.ui.readInput()))) {}
```
Perhaps you can consider
1. naming boolean b with a variable name like `isRunning` to be more descriptive
2. write the code for reading input and parsing commands on multiple lines to aid readability

Let me know what you think!

jiayi1129 reviewed (commented):
Your code is concise and precise and does its job well. Perhaps you can consider adding more comments and also going through the feedback to write code with greater readability! Still, good job thus far! 👍

Commit:
added data/ folder to .gitignore

Commit:
removed data from commit

Commit:
A-MoreOOP: Use Even More OOP

Commit:
A-JUnit: Add JUnit Tests

Commit:
A-Jar: Create a JAR File

Commit:
A-JavaDoc: JavaDoc

Commit:
Level-9: Find

Commit:
Level-9: Find

Commit:
added gradle files

Commit:
reloaded gradle projects

Commit:
GUI

Commit:
Add assertions to Pair class constructor

Pair class constructor allows the parameters Command or List to be Null.

This may cause undefined behavior in the program.

Let's add assertions to ensure Pair cannot be constructed with null parameters.

Preventing null parameters will eliminate undefined behaviors stemming from the Pair class.

davelowqx committed:
Merge pull request #1 from davelowqx/branch-A-Assertions

Add assertions to Pair class constructor

Commit:
Separate user interface into new package Duke.Ui

Classes for the user interface (using JavaFX) are found in the same package as the core program logic.

The code becomes quite disorganised, with many class files in the same root folder.

Let's create a new package Duke.Ui to encapsulate all classes for the user interface.

This adds hierarchy and organisation to the project.

davelowqx committed:
Merge pull request #2 from davelowqx/branch-A-CodeQuality

Separate user interface into new package Duke.Ui

Commit:
Add function prototypes for snooze feature

davelowqx committed:
Add snooze feature

Commit:
Add user guide and screenshot

Commit:
Add checkstyle.xml

Commit:
Add launcher to workaround classpath issues
